File: backend/app/api/v1/transactions.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.models.transaction import Transaction
from app.ml.predict import predict_transaction
from app.api.v1.auth import get_current_user

router = APIRouter(
    prefix="/transactions",
    tags=["Transactions"],
)

# =========================================================
# Get all transactions
# =========================================================

@router.get("/")
def get_transactions(
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    transactions = (
        db.query(Transaction)
        .order_by(Transaction.id)
        .all()
    )

    return transactions
# =========================================================
# Get single transaction
# =========================================================

@router.get("/{transaction_id}")
def get_transaction(
    transaction_id: int,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    transaction = (
        db.query(Transaction)
        .filter(Transaction.id == transaction_id)
        .first()
    )

    if transaction is None:
        raise HTTPException(
            status_code=404,
            detail="Transaction not found",
        )

    return transaction


# =========================================================
# Get transaction abuse risk
# =========================================================

@router.get("/{transaction_id}/risk")
def get_transaction_risk(
    transaction_id: int,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    """
    Calculate ML-based abuse risk for a transaction.
    """

    # -----------------------------------------------------
    # Find transaction
    # -----------------------------------------------------

    transaction = (
        db.query(Transaction)
        .filter(
            Transaction.id == transaction_id
        )
        .first()
    )

    if transaction is None:
        raise HTTPException(
            status_code=404,
            detail="Transaction not found.",
        )
    

    # -----------------------------------------------------
    # Predict abuse probability
    # -----------------------------------------------------

    risk_score = predict_transaction(
        db,
        transaction,
    )

    '''

    # -----------------------------------------------------
    # Convert probability into classification
    # -----------------------------------------------------

    if risk_score >= 0.50:
        prediction = "ABUSE"
    else:
        prediction = "NORMAL"
    '''

    # -----------------------------------------------------
    # Return API response
    # -----------------------------------------------------

    return {
        "transaction_id": transaction.id,
        "transaction_code": transaction.transaction_code,
        "risk_score": risk_score,
        "risk_percentage": round(
            risk_score * 100,
            2,
        ),
        "prediction": (
           "ABUSE"
            if risk_score >= 0.5
            else "NORMAL" 
        ),
    }

File: frontend/src/api/transactions.ts
import { apiRequest } from "./client";

export interface Transaction {
  id: number;
  transaction_code: string;
  amount: number;
  currency?: string;
  is_abuse: boolean;
}

export function getTransactions(): Promise<Transaction[]> {
  const token = localStorage.getItem("access_token");

  return apiRequest<Transaction[]>("/transactions/", {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
}

export function getTransaction(
  transactionId: number
): Promise<Transaction> {
  const token = localStorage.getItem("access_token");

  return apiRequest<Transaction>(
    `/transactions/${transactionId}`,
    {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    }
  );
}

export interface TransactionRisk {
  transaction_id: number;
  transaction_code: string;
  risk_score: number;
  risk_percentage: number;
  prediction: "ABUSE" | "NORMAL";
}

export function getTransactionRisk(
  transactionId: number
): Promise<TransactionRisk> {
  const token = localStorage.getItem("access_token");

  return apiRequest<TransactionRisk>(
    `/transactions/${transactionId}/risk`,
    {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    }
  );
}